### PR TITLE
Adding explain option in mlflow deployments cli

### PR DESCRIPTION
**Pull request by shrinath-suresh**
Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

## What changes are proposed in this pull request?

https://pytorch.org/serve/inference_api.html#explanations-api 

To support torchserve explain api (Captum) , adding a new deployment cli option `explain`.

Note: the option is only available in cli mode. Explain method has not been added to `BaseDeploymentClient` as it would affect other flavours. 

## How is this patch tested?

Added necessary unit tests

https://github.com/chauhang/mlflow-torchserve-1/pull/7 - Separate activity going on to add few examples with `mlflow explain cli` commands. 

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [x] `language/python`: Python APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py [hunk 1]
```diff
@@ -27,6 +27,10 @@ def get_deployment(self, name):
     def predict(self, deployment_name, df):
         return "1"
 
+    def explain(self, deployment_name, df):
+        # pylint: disable=unused-argument
```
Comment (shrinath-suresh) [inline tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py:31]: Since explain method is not availabe in basedeployment client, pylint throws unused argument warning. Ignoring the warning.

File: mlflow/deployments/cli.py [hunk 2]
```diff
@@ -258,3 +258,24 @@ def predict(target, name, input_path, output_path):
...
+@parse_input
+@parse_output
+def explain(target, name, input_path, output_path):
+    """
+    predict the features importance for the deployed model for the given input(s)
+    """
+    import pandas as pd
+
+    df = pd.read_json(input_path)
+    client = interface.get_deploy_client(target)
```
Comment (shrinath-suresh) [inline mlflow/deployments/cli.py:275]: Just a thought:  Should we raise an exception here if target is not `torchserve` ?

Comment (smurching) [reply mlflow/deployments/cli.py:275]: @shrinath-suresh we probably don't want to condition logic in MLflow core on specific plugins that exist outside MLflow (you could imagine then we'd have a big branching if-statement for all deploy plugins/in order to add a new plugin folks would have to know to merge a PR modifying this file with details of their own plugin, vs being able to develop & release a plugin independently of MLflow.

Could we instead check whether the `explain` method is defined on the `client` object and raise an exception saying that the current plugin doesn't support the `explain` API if not? e.g. something like

```suggestion
    client = interface.get_deploy_client(target)
     if not hasattr(client, "explain"):
         raise MlflowException("Computing model explanations is not yet supported for deployment target %s" % target)
     result = client.explain(name, df)
     ... 
```

File: mlflow/deployments/cli.py [hunk 3]
```diff
@@ -258,3 +258,24 @@ def predict(target, name, input_path, output_path):
...
+
+
+@commands.command("explain")
+@deployment_name
+@target_details
+@parse_input
+@parse_output
+def explain(target, name, input_path, output_path):
+    """
+    predict the features importance for the deployed model for the given input(s)
```
Comment (smurching) [inline mlflow/deployments/cli.py:270]: Could we enhance this docstring to make it clearer what the API does? Here's a stab at a docstring, let me know if I'm understanding the purpose of this API correctly and thanks in advance!:

```suggestion
    """
    Generate explanations of model predictions on the specified input for the deployed model for the given input(s). Explanation output formats vary by deployment target, and can include details like feature importance for understanding/debugging predictions. Run `mlflow deployments help` or consult the documentation for your plugin for details on explanation format. For information about the input data formats accepted by this function, see the following documentation: https://www.mlflow.org/docs/latest/models.html#built-in-deployment-tools
    """    
```

@shrinath-suresh it'd be good to check - given this explanation, could a new user successfully understand how to generate/interpret model explanations from a model deployed to torchserve?

File: tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py (hunk 1 above)
Comment (smurching) [reply tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py:31]: I think if we want to add this API to the MLflow core CLI, we'll need to add it to the base deployment client - it's a bit strange from a user perspective for this functionality to be CLI-only, as it's then not possible to write e.g. a Python script to perform deployment and all other deploy plugin functionality is available via both the API and CLI

Comment (KasirajanA) [reply tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py:31]: If we declare the explain method as abstract method in base deployment class then client instance creation for other flavour's will fail if the abstract method is not implemented in those plugin classes. Please refer to the attached snapshots. Can we add explain as a non-abstract method in base deployment class so that it won't break the other flavour's instance creation?

explain method as abstract method.
Screenshot from 2021-05-31 17-43-29

explain as non-abstract method
Screenshot from 2021-05-31 17-45-57

File: mlflow/deployments/cli.py (hunk 2 above)
Comment (KasirajanA) [reply mlflow/deployments/cli.py:275]: If we add explain method as non-abstract method in base deployment class then "hasattr(client, "explain")" will always return True. So, Can we add the exception in the base class's explain method itself?

```
    def explain(self, deployment_name, df):
        raise MlflowException("Computing model explanations is not yet supported for this deployment target")
```

Comment (smurching) [reply mlflow/deployments/cli.py:275]: @KasirajanA yep that makes sense!

Comment (smurching) [reply mlflow/deployments/cli.py:275]: And I think adding it as a non-abstract method (good catch there as well) should allow us to make this change without breaking existing plugins, as you described :)

**Comment by smurching** [comment]
BTW, thanks for this PR - per offline discussion, it seems reasonable to me to add an `explain` API as an optional method for plugins to implement (raising a clear exception indicating that explanation support isn't yet supported). In particular, it seems possible to implement explanation support for our existing built-in sagemaker and azureml deployment targets:
* [Sagemaker explanations doc link](https://sagemaker-examples.readthedocs.io/en/latest/sagemaker_processing/fairness_and_explainability/fairness_and_explainability.html#Detecting-Bias) 
* [AzureML explanations doc link](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-machine-learning-interpretability-aml#interpretability-at-inference-time)

In particular:
* With Sagemaker, the workflow is that Sagemaker will spin up additional instances to compute explanations on a deployed model. As such the API seems similar to the one proposed here, with a few additional required parameters - you need to additionally specify an instance type and instance count for the compute launched for explanations (specified as part of a [ModelConfig](https://sagemaker.readthedocs.io/en/stable/api/training/processing.html#sagemaker.clarify.ModelConfig)), the desired input/output S3 path of the data/explanations (should be doable to convert this into local file paths in the MLflow case by uploading/downloading from a default, optionally user-specified bucket), and label/feature columns (can hopefully be inferred from the model schema in the MLflow case, or otherwise specified by the user)
* With AzureML, it seems you deploy an explainer alongside your main ML model (i.e. you pass both an explainer & your ML model to the `Webservice.deploy_from_model` API as in step f (at the time of writing) of [this example](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-machine-learning-interpretability-aml#interpretability-at-inference-time)). We should be able to implement computing explanations by updating the Webservice to include an explainer ([see API](https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.webservice.aci.aciwebservice?view=azure-ml-py#azureml_core_webservice_aci_AciWebservice_update)), optionally adding support for specifying an explainer at deploy time & later computing explanations against it.

**Comment by KasirajanA** [comment]
@smurching @harupy, Can you please approve the workflows to run?

**Comment by akshaya-a** [comment]
@smurching  ack, pointing the team at this PR

**Comment by smurching** [comment]
@KasirajanA sorry for the delay, just triggered a CI run :)

File: tests/pytorch/test_pytorch_model_export.py [hunk 4]
```diff
@@ -807,6 +808,11 @@ def test_load_model_succeeds_when_data_is_model_file_instead_of_directory(
 
 
 @pytest.mark.large
+@pytest.mark.skipif(
+    Version(torch.__version__) >= Version("1.9.0"),
+    reason="From pytorch 1.9.0, torch.load() creates a subclass from the passed Unpickler. "
+           "Refer https://github.com/pytorch/pytorch/pull/53139",
+)
```
Comment (KasirajanA) [inline tests/pytorch/test_pytorch_model_export.py:815]: In pytorch 1.9.0, a subclass is created from the passed Unpickler in the torch.load function as pickle.Unpickler is marked readonly in pickle.
For more details refer to [this PR](https://github.com/pytorch/pytorch/pull/53139)

File: mlflow/deployments/base.py [hunk 5]
```diff
@@ -190,3 +191,20 @@ def predict(self, deployment_name, df):
...
         pass
+
+    @experimental
+    def explain(self, deployment_name, df):  # pylint: disable=unused-argument
+        """
+        Generate explanations of model predictions on the specified input pandas Dataframe
+        ``df`` for the deployed model. Explanation output formats vary by deployment target,
+        and can include details like feature importance for understanding/debugging predictions.
+        Note that we have created this function as non-abstract method to not break
+        the existing downstream.
```
Comment (smurching) [inline mlflow/deployments/base.py:202]:
```suggestion
```

Thanks for the updated docstring! We should remove these lines IMO, I don't think we want this to show up in user-facing docs (e.g. if I'm using the deploy API I might be confused by developer-facing comments about this method being non-abstract)

File: mlflow/deployments/base.py [hunk 6]
```diff
@@ -190,3 +191,20 @@ def predict(self, deployment_name, df):
...
+
+        :param deployment_name: Name of deployment to predict against
+        :param df: Pandas DataFrame to use for explaining feature importance in model prediction
+        :return: An exception if the implementation is not available in deployment target's class
```
Comment (smurching) [inline mlflow/deployments/base.py:206]: Could we clarify that the expected return type is a JSON-able object, e.g. a pandas dataframe or numpy array? 
```suggestion
        :return: A JSON-able object (pandas dataframe, numpy array, dictionary), or exception if unimplemented for the current deployment target
```

File: mlflow/deployments/cli.py [hunk 7]
```diff
@@ -258,3 +258,31 @@ def predict(target, name, input_path, output_path):
...
+
+@commands.command("explain")
+@deployment_name
+@target_details
+@parse_input
+@parse_output
+def explain(target, name, input_path, output_path):
+    """
+    Generate explanations of model predictions on the specified input for
+    the deployed model for the given input(s).Explanation output formats vary
```
Comment (smurching) [inline mlflow/deployments/cli.py:271]:
```suggestion
    the deployed model for the given input(s). Explanation output formats vary
```

**Review by smurching — COMMENTED** [review]
Mostly looks great! Just a few minor doc comments, otherwise looks good. Were you able to manually QA the new deployments API using both `mlflow` and the skinny client? @harupy may be able to provide guidance on how to do so. Thanks again for the PR :)

**Comment by shrinath-suresh** [comment]
> Mostly looks great! Just a few minor doc comments, otherwise looks good. Were you able to manually QA the new deployments API using both `mlflow` and the skinny client? @harupy may be able to provide guidance on how to do so. Thanks again for the PR :)

Thanks for the review sid. With this branch, we have tested the api behaviour on `mlflow-torchserve` plugin side. @harupy can you please point us to documentation on how to test it in `mlflow` and the skinny client.

**Comment by harupy** [comment]
For skinny-client, I think you can install it via: `pip install mlflow-skinny` and verify the CLI commands work as expected.

**Comment by KasirajanA** [comment]
@smurching , @harupy, we installed mlflow-skinny and ran the pytorch examples successfully. 
mlflow_skinny

**Comment by shrinath-suresh** [comment]
> @smurching , @harupy, we installed mlflow-skinny and ran the pytorch examples successfully.
> mlflow_skinny

@KasirajanA Can you please test the cli commands as well (create, predict and explain) ?

**Review by smurching — APPROVED** [review]
LGTM! If manual QA pans out then let's merge :)

**Comment by shrinath-suresh** [comment]
@KasirajanA  has tested the example with explain API and its working as expected.

titanic_explain

**Comment by shrinath-suresh** [comment]
@harupy @smurching Shall we go ahead and merge this PR ?
